subsys/shell/shell.c

```diff
@@ -1,652 +1,1696 @@
+
+	while (length) {
+		int err = shell->iface->api->write(shell->iface,
+				&((u8_t const *) data)[offset], length,
+				&tmp_cnt);
+		(void)err;
+		assert(err == 0);
+		assert(length >= tmp_cnt);
+		offset += tmp_cnt;
+		length -= tmp_cnt;
+		if (tmp_cnt == 0 &&
+		    (shell->ctx->state != SHELL_STATE_PANIC_MODE_ACTIVE)) {
+			/* todo  semaphore pend*/
+			if (IS_ENABLED(CONFIG_MULTITHREADING)) {
+				k_poll(&shell->ctx->events[SHELL_SIGNAL_TXDONE],
+				1, K_FOREVER);
+			} else {
+				/* Blocking wait in case of bare metal. */
+				while (!shell->ctx->internal.flags.tx_rdy) {
 
-		str = app_prompt_handler();
-		if (str) {
-			return str;
+				}
+				shell->ctx->internal.flags.tx_rdy = 0;
+			}
 		}
 	}
+}
 
-	if (default_module) {
-		if (default_module->prompt) {
-			const char *ret;
+/* @brief Function shall be used to search commands.
+ *
+ * It moves the pointer entry to command of static command structure. If the
+ * command cannot be found, the function will set entry to NULL.
+ *
+ *   @param command	Pointer to command which will be processed (no matter
+ *			the root command).
+ *   @param lvl		Level of the requested command.
+ *   @param idx		Index of the requested command.
+ *   @param entry	Pointer which points to subcommand[idx] after function
+ *			execution.
+ *   @param st_entry	Pointer to the structure where dynamic entry data can be
+ *			stored.
+ */
+static void cmd_get(const struct shell_cmd_entry *command, size_t lvl,
+		    size_t idx, const struct shell_static_entry **entry,
+		    struct shell_static_entry *d_entry)
+{
+	assert(entry != NULL);
+	assert(command != NULL);
+	assert(d_entry != NULL);
 
-			ret = default_module->prompt();
-			if (ret) {
-				return ret;
-			}
+	if (lvl == SHELL_CMD_ROOT_LVL) {
+		if (idx < shell_root_cmd_count()) {
+			const struct shell_cmd_entry *cmd;
+
+			cmd = shell_root_cmd_get(idx);
+			*entry = cmd->u.entry;
+		} else {
+			*entry = NULL;
 		}
+		return;
+	}
 
-		return default_module_prompt;
+	if (command == NULL) {
+		*entry = NULL;
+		return;
 	}
 
-	return prompt;
+	if (command->is_dynamic) {
+		command->u.dynamic_get(idx, d_entry);
+		*entry = (d_entry->syntax != NULL) ? d_entry : NULL;
+	} else {
+		*entry = (command->u.entry[idx].syntax != NULL) ?
+				&command->u.entry[idx] : NULL;
+	}
 }
 
-static void line_queue_init(void)
+static void vt100_color_set(const struct shell *shell,
+			    enum shell_vt100_color color)
 {
-	int i;
 
-	for (i = 0; i < MAX_CMD_QUEUED; i++) {
-		k_fifo_put(&avail_queue, &buf[i]);
+	if (shell->ctx->vt100_ctx.col.col == color) {
+		return;
 	}
-}
 
-static size_t line2argv(char *str, char *argv[], size_t size)
-{
-	size_t argc = 0;
+	shell->ctx->vt100_ctx.col.col = color;
 
-	if (!strlen(str)) {
-		return 0;
-	}
+	if (color != SHELL_DEFAULT) {
 
-	while (*str && *str == ' ') {
-		str++;
+		u8_t cmd[] = SHELL_VT100_COLOR(color - 1);
+
+		shell_raw_fprintf(shell->fprintf_ctx, "%s", cmd);
+	} else {
+		static u8_t const cmd[] = SHELL_VT100_MODESOFF;
```

Comment:
Please use "static const" consistently for such cases. That's how Zephyr codebase has it, and that allows easily search for such cases.

Reply:
Also, maybe define SHELL_VT100_MODESOFF and other const sequences as static const u8_t[] altogether and use directly (instead of preprocessor define of array initializer) ?